File: CH1_Stack_Queue/1_getMin.cpp
/*
2019-9.24
CH1 栈--设计一个有getMin功能的栈，测试用例
6
push 3
push 2
push 1
getMin
pop
getMin
*/
#include <iostream>
#include <cstdio>
#include <stack>
#include <string>
using namespace std;

int main()
{
	stack<int> stackData, stackMin;
	// freopen("in1.txt", "r", stdin);     
	int op;
	cin >> op;
	for (int i = 0; i < op; i++)
	{
		string s;
		cin >> s;
		if (s == "push")
		{
			int num;
			cin >> num;
			stackData.push(num);    // 栈不为空时stackMin.top()会出现段错误
			if (stackMin.empty() || num <= stackMin.top())
				stackMin.push(num);
		}
		else if (s == "pop")
		{
			int num = stackData.top();
			stackData.pop();
			if (num == stackMin.top())
				stackMin.pop();
		}
		else
		{
			cout << stackMin.top() << endl;
		}
	}	
	return 0;
}

File: CH1_Stack_Queue/1_getMin.py
'''
2019-10.25
CH1 栈--设计一个有getMin功能的栈，测试用例
6
push 3
push 2
push 1
getMin
pop
getMin
'''
import pdb


n = int(input().strip())
stack = []
stack_min = []
for _ in range(n):
	x = input().split()
	pdb.set_trace()

	# print(x)
	if x[0] == "push":
		num = int(x[1])      # 大坑--必须进行int()类型转换,否则存入的是字符串,负数时有bug
		stack.append(num)
		if stack_min == [] or stack_min[-1] >= num:
			stack_min.append(num)
		else:
			stack_min.append(stack_min[-1])
	elif x[0] == "pop":
		stack_min.pop()
		stack.pop()
	else:
		print(stack_min[-1])


